File: 23-Graph/python/10_number_of_islands.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        rows, cols = len(grid), len(grid[0])
        parent = {}
        rank = {}
        self.count = 0

        def find(i):
            if parent[i] != i:
                parent[i] = find(parent[i])
            return parent[i]

        def union(i, j):
            root_i = find(i)
            root_j = find(j)
            if root_i != root_j:
                if rank[root_i] < rank[root_j]:
                    parent[root_i] = root_j
                elif rank[root_i] > rank[root_j]:
                    parent[root_j] = root_i
                else:
                    parent[root_j] = root_i
                    rank[root_i] += 1
                self.count -= 1

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    cell_id = r * cols + c
                    parent[cell_id] = cell_id
                    rank[cell_id] = 0
                    self.count += 1

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    cell_id = r * cols + c
                    if c + 1 < cols and grid[r][c + 1] == '1':
                        union(cell_id, r * cols + c + 1)
                    if r + 1 < rows and grid[r + 1][c] == '1':
                        union(cell_id, (r + 1) * cols + c)

        return self.count

File: 23-Graph/cpp/10-number-of-islands.cpp
#include <vector>

using namespace std;

class Solution {
private:
    vector<int> parent;
    vector<int> rank;
    int count;

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void unite(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX == rootY) return;

        if (rank[rootX] < rank[rootY]) {
            parent[rootX] = rootY;
        } else if (rank[rootX] > rank[rootY]) {
            parent[rootY] = rootX;
        } else {
            parent[rootY] = rootX;
            rank[rootX]++;
        }
        count--;
    }

public:
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();
        parent.resize(rows * cols);
        rank.resize(rows * cols, 0);
        count = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    int id = i * cols + j;
                    parent[id] = id;
                    count++;
                }
            }
        }

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    int id = i * cols + j;
                    if (j + 1 < cols && grid[i][j + 1] == '1') {
                        unite(id, i * cols + j + 1);
                    }
                    if (i + 1 < rows && grid[i + 1][j] == '1') {
                        unite(id, (i + 1) * cols + j);
                    }
                }
            }
        }

        return count;
    }
};
